lib/chef/provider/package/yum.rb

```diff
@@ -153,6 +153,41 @@ def unlock_package(names, versions)
 
         private
 
+        def target_version(package_name, new_version, current_version, candidate_version)
+          return super if action != :install || current_version.nil? || new_version
```

Comment:
couldn't you run into the case where you have a x86_64 pkg installed at X, but you're calling `:upgrade` on on the i386 pkg whose candidate version is Y (where Y > X), and thus need the logic below?

Reply:
I do not believe there is a legitimate way for that to occur.  In the scenario where Y is the latest candidate for an i386 package, there must also be an x86_64 candidate at version Y, otherwise the repository is incomplete and could never install version Y of the i386 package.  I believe performing a bare `yum upgrade` in a on a system with a repository in such a state would fail unless `--skip-broken` was used.

Of course, when the action is `:upgrade` with the x86_64 package at version X and candidate Y is in the repository, `super` will display a message about updating the x86_64 pacakge to version Y, and another message about installing or updating to version Y of the i386 package, as appropriate.

Reply:
OK so, current state of system:
* foo.x86_64 installed at 1.2-1
* foo.i386 not installed
Recipe:
* `package 'foo.i386' { :upgrade }`
Repo:
* 1.2-1 and 1.3-1 are both available

I believe this is the case where you need the logic below and the return here will break things. I think.

Reply:
If by "break things" you mean fall back to existing behavior, that is correct.  These modifications only change the behavior of the `:install` action, not the `:upgrade` action.  That is because when you use `:install` and don't specify a version, you are telling `chef` that you only care that the package gets installed, in which case it would be acceptable for `chef` to choose whichever version it determines is most appropriate.  When you use the `:upgrade` action, you are telling `chef` to ensure the package is at the latest version.  It seems more appropriate to have `chef` continue to generate an error when `:upgrade` is used for `foo.i386` and version 1.3-1 is available than to install an out of date version of `foo.i386`.

The alternative would be to try to have `chef` implicitly upgrade `foo.x86_64` every time there is an update for `foo.i386` so that it could complete the operation successfully.  However, in this case, `chef` would be modifying `foo.x86_64`, which it was never told to manage.  I would not expect this behavior and would view it as a side effect.  I believe it would be better for the recipe to be more explicit.  Specifically,
```ruby
package %i[foo.i686 foo.x86_64] do
    action :upgrade
end
```
behaves as expected without this patch, whereas `action :install` does not.

It is also worth noting that the provider only knows about the packages it was told to mange.  In your scenario, the provider does not know whether `foo.x86_64` is present.  Even though the helper script has access to this information, the provider doesn't have a way to ask for it.  We would have to make significant changes to the provider  and helper to allow the provider to implicitly identify and manage packages it was not instructed to manage.  Given that I could find no tests for this provider and I rarely see the `:upgrade` action used in the wild, I think refactoring the code to add such a feature could introduce more risk than the benefits merit.